### PR TITLE
Retry mechanism

### PR DESCRIPTION
**Pull request by GunaKKIBM**
This is a fix added for the following bug:
```
https://github.ibm.com/alchemy-containers/customer-tickets/issues/15931
```

Summary:
1. Token exchange URL is first fetched from cloud-conf config map. If this config map doesn't exist:
- For VPC Gen2 cluster - always private IAM URL is used.
- For others - Classic/IPI/Satellite - user provided URL is used.
2. Retry happens upto 5 mins (exponentially retryGap increases) whenever a timeout occurs while reaching out to token exchange URL.
3. On a VPC Gen2 cluster, if private IAM endpoint is unreachable even after retrying for 5 mins, we switch to public IAM endpoint and try to fetch the token. But, after that, endpoint is set back to private IAM endpoint.

### CODE REVIEW

File: pkg/authenticator/authenticator.go [hunk 1]
```diff
@@ -28,8 +29,17 @@ import (
...
 	ProviderType string = "ProviderType"
-	SecretKey    string = "SecretKey"
+
+	// SecretKey ...
+	SecretKey string = "SecretKey"
+
+	// maxRetryAttempt ...
+	maxRetryAttempt = 15
+
+	// maxRetryGap ...
```
Comment (skorati) [inline pkg/authenticator/authenticator.go:41]: Is there any reason for using the retrygap as  8, 4 ?
4 -> 1st attempt
8 -> 14 attempts 
Total of 116 seconds?

File: pkg/authenticator/authenticator.go [hunk 2]
```diff
@@ -233,3 +243,38 @@ func resetIAMURL(auth Authenticator) bool {
...
+		}
+
+		time.Sleep(time.Second * time.Duration(retryGap))
+		retryGap = retryGap * 2
+		if retryGap > maxRetryGap {
+			retryGap = maxRetryGap
+		}
+	}
+
+	// Reset the IAM URL from to public, if it is private
```
Comment (skorati) [inline pkg/authenticator/authenticator.go:272]: Typo
// Reset the IAM URL to public, if it is private
// Retry fetching IAM token again

**Review by skorati — COMMENTED** [review]
Added comments

File: pkg/authenticator/authenticator.go (hunk 1 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go:41]: Yes. will discuss with team regarding the ideal number. My thought was, retrying for 2 mins. We can try upto 5 mins. We can discuss on this

File: pkg/authenticator/authenticator.go (hunk 2 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go]: Updated

File: pkg/authenticator/authenticator.go [hunk 3]
```diff
@@ -233,3 +243,38 @@ func resetIAMURL(auth Authenticator) bool {
...
+		retryGap = retryGap * 2
+		if retryGap > maxRetryGap {
+			retryGap = maxRetryGap
+		}
+	}
+
+	// Reset the IAM URL to public, if it is private
+	// Retry fetching IAM token again
+	if resetIAMURL(authenticator) {
+		logger.Info("Updated IAM URL from private to public, retrying to fetch IAM token")
```
Comment (GunaKKIBM) [inline pkg/authenticator/authenticator.go:275]: Move out resetURL to the caller function

File: pkg/authenticator/authenticator.go [hunk 4]
```diff
@@ -233,3 +243,38 @@ func resetIAMURL(auth Authenticator) bool {
...
+		time.Sleep(time.Second * time.Duration(retryGap))
+		retryGap = retryGap * 2
+		if retryGap > maxRetryGap {
+			retryGap = maxRetryGap
+		}
+	}
+
+	// Reset the IAM URL to public, if it is private
+	// Retry fetching IAM token again
+	if resetIAMURL(authenticator) {
```
Comment (arahamad) [inline pkg/authenticator/authenticator.go:274]: Can you move this out of this method

**Review by arahamad — CHANGES_REQUESTED** [review]

File: pkg/authenticator/authenticator.go (hunk 1 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go:41]: Upto 1 min, keep exponential retry. 2, 4., 8, 16 ...

File: pkg/authenticator/authenticator.go [hunk 5]
```diff
@@ -233,3 +243,38 @@ func resetIAMURL(auth Authenticator) bool {
...
+	var err error
+
+	for retryAttempt := 0; retryAttempt < maxRetryAttempt; retryAttempt++ {
+		err := retryfunc()
+		if err == nil {
+			return err
+		}
+
+		logger.Error("Error fetching fresh token", zap.Error(err), zap.Int("AttemptNo", retryAttempt+1))
+		// return if the error is not timeout
```
Comment (GunaKKIBM) [inline pkg/authenticator/authenticator.go:260]: Elaborate the comment

File: pkg/authenticator/api-key.go [hunk 6]
```diff
@@ -54,20 +54,26 @@ func (aa *APIKeyAuthenticator) GetToken(freshTokenRequired bool) (string, uint64
...
-		if !isTimeout(err) {
+		// Reset the IAM URL to public, if it is private
+		if !resetIAMURL(aa) {
 			return "", tokenlifetime, utils.Error{Description: "Error fetching iam token using api key", BackendError: err.Error()}
 		}
-
-		aa.logger.Info("Updating iam URL to public, if it is private and retrying to fetch token")
-		if !resetIAMURL(aa) {
+		// Retry fetching IAM token again
+		aa.logger.Info("Updated IAM URL from private to public, retrying to fetch IAM token")
```
Comment (skorati) [inline pkg/authenticator/api-key.go:69]: can we rewrite this error message? 
retrying token fetch, after switching from private to public IAM URL..

File: pkg/authenticator/authenticator.go [hunk 7]
```diff
@@ -233,3 +243,40 @@ func resetIAMURL(auth Authenticator) bool {
 	}
 	return false
 }
+
+// retry ...
+func retry(logger *zap.Logger, retryfunc func() error) error {
+
```
Comment (skorati) [inline pkg/authenticator/authenticator.go:249]: Could you please add  a TODO here for making maxretry attempts/gap as configurable.

**Review by skorati — APPROVED** [review]
Looks good, 
I added minor comments please have a look.

File: pkg/authenticator/api-key.go [hunk 8]
```diff
@@ -54,20 +54,26 @@ func (aa *APIKeyAuthenticator) GetToken(freshTokenRequired bool) (string, uint64
...
+		tokenResponse, err = aa.authenticator.RequestToken()
+		return err
+	})
+
 	if err != nil {
-		aa.logger.Error("Error fetching fresh token", zap.Error(err))
-		// If the cluster cannot access private iam endpoint, hence returns timeout error, switch to public IAM endpoint.
-		if !isTimeout(err) {
+		// Reset the IAM URL to public, if it is private
+		if !resetIAMURL(aa) {
```
Comment (arahamad) [inline pkg/authenticator/api-key.go:65]: when reset is required ?

Comment (GunaKKIBM) [reply pkg/authenticator/api-key.go:70]: When timeout error still exists using private endpoint.

File: pkg/authenticator/api-key.go (hunk 6 above)
Comment (GunaKKIBM) [reply pkg/authenticator/api-key.go:75]: Done

File: pkg/authenticator/authenticator.go (hunk 7 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go:249]: Done

File: pkg/authenticator/pod-identity.go [hunk 9]
```diff
@@ -58,19 +59,33 @@ func (ca *ComputeIdentityAuthenticator) GetToken(freshTokenRequired bool) (strin
...
+		// If the error is not related to timeout or if the token exchange URL is provided by user, return error.
+		if !isTimeout(err) || ca.userProvidedURL {
+			return "", tokenlifetime, utils.Error{Description: "Error fetching iam token using trusted profile", BackendError: err.Error()}
 		}
 
-		ca.logger.Info("Updating iam URL to public, if it is private and retrying to fetch token")
-		if !resetIAMURL(ca) {
-			return "", tokenlifetime, utils.Error{Description: "Error fetching iam token using compute identity", BackendError: err.Error()}
+		// By default authenticator uses private IAM URL, setting it to public
+		setPrivateIAMURL(ca)
```
Comment (GunaKKIBM) [inline pkg/authenticator/pod-identity.go:75]: set public IAM URL

File: pkg/config/cluster_config.go [hunk 10]
```diff
@@ -64,40 +65,51 @@ func GetClusterInfo(kc k8s_utils.KubernetesClient, logger *zap.Logger) (ClusterC
...
-	return FrameTokenExchangeURLFromClusterInfo(isSatellite, cc, logger)
+	return FrameTokenExchangeURLFromClusterInfo(clusterInfo, logger)
 }
 
 // GetTokenExchangeURLfromStorageSecretStore ...
-func GetTokenExchangeURLfromStorageSecretStore(isSatellite bool, config Config, providerType string) (string, error) {
+func GetTokenExchangeURLfromStorageSecretStore(clusterInfo ClusterConfig, config Config, providerType string) (string, bool, error) {
+
+	// Possible conditions and env
+	// Return Private Prod/Stage IAM URL if the cluster is VPC Gen2
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:102]: Update the comment.

File: pkg/config/cluster_config.go [hunk 11]
```diff
@@ -109,37 +121,30 @@ func GetTokenExchangeURLfromStorageSecretStore(isSatellite bool, config Config,
...
+func FrameTokenExchangeURLFromClusterInfo(cc ClusterConfig, logger *zap.Logger) (string, bool) {
 
-	if !strings.Contains(cc.MasterURL, stageMasterURLsubstr) {
+	var providedTokenExchangeURL = true
+	isSatellite := IsSatellite(cc, logger)
+	if isProduction(cc.MasterURL) {
 		logger.Info("Env-Production")
 		if isSatellite {
-			return (utils.ProdPublicIAMURL + tokenExchangePath)
+			return (utils.ProdPublicIAMURL + tokenExchangePath), providedTokenExchangeURL
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:138]: Consolidate these calls to one function

File: pkg/config/cluster_config.go [hunk 12]
```diff
@@ -64,40 +65,51 @@ func GetClusterInfo(kc k8s_utils.KubernetesClient, logger *zap.Logger) (ClusterC
...
 }
 
 // GetTokenExchangeURLfromStorageSecretStore ...
-func GetTokenExchangeURLfromStorageSecretStore(isSatellite bool, config Config, providerType string) (string, error) {
+func GetTokenExchangeURLfromStorageSecretStore(clusterInfo ClusterConfig, config Config, providerType string) (string, bool, error) {
+
+	// Possible conditions and env
+	// Return Private Prod/Stage IAM URL if the cluster is VPC Gen2
+	var providedTokenExchangeURL = false
+	if clusterInfo.ClusterType == utils.VPCGen2 {
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:104]: Make it GetIAASProvider

File: pkg/config/cluster_config.go [hunk 13]
```diff
@@ -158,3 +162,8 @@ func IsSatellite(cc ClusterConfig, logger *zap.Logger) bool {
 
 	return false
 }
+
+// GetIAASProvider ...
+func GetIAASProvider(cc ClusterConfig) string {
+	return cc.ClusterType
```
Comment (GunaKKIBM) [inline pkg/config/cluster_config.go:168]: Add comments on why this was segregated to a separate method.

File: pkg/config/cluster_config.go [hunk 14]
```diff
@@ -109,37 +120,30 @@ func GetTokenExchangeURLfromStorageSecretStore(isSatellite bool, config Config,
...
-		return url, nil
-	}
-
-	isProd := isProduction(url)
-	if isProd {
-		return utils.ProdPrivateIAMURL + tokenExchangePath, nil
-	}
-	return utils.StagePrivateIAMURL + tokenExchangePath, nil
+	return url, providedTokenExchangeURL, nil
 }
```
Comment (skorati) [inline pkg/config/cluster_config.go:128]: @GunaKKIBM could you please some documentation to this switch case.

File: pkg/utils/constants.go [hunk 15]
```diff
@@ -57,4 +57,10 @@ const (
 	Softlayer = "softlayer"
 	// SatelliteProvider ...
 	SatelliteProvider = "upi"
+	// VPCGen2 refers to VPC cluster type ...
```
Comment (skorati) [inline pkg/utils/constants.go:60]: @GunaKKIBM in the last review meeting we had some confusion on terms w.r.t to infrastructure vs cluster.
I hope the below constants are for cluster, I mean `vpc-gen2_cruiser` means the cluster is of type vpc and `satellite_cruiser` its satellite cluster irrespective of the infrasture?

**Review by skorati — APPROVED** [review]
looks good,
just some minor suggestions.

File: pkg/authenticator/authenticator.go (hunk 4 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go]: Done

File: pkg/authenticator/authenticator.go (hunk 5 above)
Comment (GunaKKIBM) [reply pkg/authenticator/authenticator.go]: Done

**Review by arahamad — APPROVED** [review]
lgtm, please verify these changes as well

**Comment by GunaKKIBM** [comment]
[Testing-private.tar.gz](https://github.com/IBM/secret-utils-lib/files/13249218/Testing-private.tar.gz) Basic test results. I have uploaded other test results internally too. 
[ROKS-public.tar.gz](https://github.com/IBM/secret-utils-lib/files/13249225/ROKS-public.tar.gz)

**Pull request merged**